### ISSUE TITLE
Numerical error with tf.nn.conv2d, tf.nn.moments and tf.sqrt

### ISSUE THREAD

**Issue by victorcampos7** [tensorflow/tensorflow#17273]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes (see code below)
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: SUSE Linux Enterprise Server 12.2 (x86_64)
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: 1.5.0 (git v1.5.0-0-g37aa430d84) & also present in 1.4.0
- **Python version**: 3.6.4
- **Bazel version (if compiling from source)**: 0.10.1
- **GCC/Compiler version (if compiling from source)**: GCC/7.2.0
- **CUDA/cuDNN version**: None
- **GPU model and memory**: None
- **Exact command to reproduce**: See script below

### Describe the problem
There is a numerical error with a very specific sequence of operations, that make tf.sqrt() return a tensor of zeros when applied on the output of tf.nn.moments (regardless of the value returned by tf.nn.moments). This error happens when the input of tf.nn.moments comes from tf.nn.conv2d, but not when removing the latter.

This error happens when running on a cluster where TensorFlow has been installed from source with "-m64 -march=native -mtune=native". We tried upgrading from 1.4.0 to 1.5.0, but the error still occurs. The following flags were used during compilation:

TF_NEED_JEMALLOC="1"
TF_NEED_HDFS="1"
TF_NEED_OPENCL="0"
TF_NEED_CUDA="0"
TF_ENABLE_XLA="0"
TF_CUDA_CLANG="0"
TF_NEED_GCP="0"
TF_NEED_MKL="1"
TF_NEED_VERBS="0"
TF_NEED_MPI="1"
TF_NEED_S3="0"
TF_NEED_GDR="0"
TF_NEED_OPENCL_SYCL="0"
TF_MKL_ROOT="[DIR]/mkl-dnn/external/mklml_lnx_2018.0.1.20171007"
export MPI_HOME="[DIR]/2018.1.038/compilers_and_libraries_2018.1.163/linux/mpi/intel64"

Android build: No.

Compiled using Intel MPI plus complete MKL (2018.1.038) loaded in the environment.

We also tested this same code in two other machines, where it works properly. One of them is an older cluster, with TensorFlow compiled from source using the same configuration; the main difference is that the processors in this cluster do not support AVX512. The other machine is a laptop where TensorFlow was installed with pip.

### Source code / logs

The smallest code to reproduce the error is the following:

```
import tensorflow as tf
import numpy as np

# TensorFlow version
print('TF version:', tf.__version__, tf.__git_version__, '\n')

# Create placeholders
o = tf.placeholder(tf.float32, [None, 84, 84, 4])

# Create graph: conv(x, w) -> (x-mean(x))/std(x)
x = o

w = tf.get_variable("w", [3, 3, x.get_shape()[-1], 16])
x = tf.nn.conv2d(x, w, [1, 1, 1, 1], padding="SAME")  # remove this and the error disappears...

avg, variance = tf.nn.moments(x, np.arange(len(x.get_shape().as_list()) - 1), keep_dims=True)
epsilon = 1e-3  # for numerical stability in sqrt

variance = tf.Print(variance, [avg], message='mean(x): ')
variance = tf.Print(variance, [variance], message='variance(x): ')
variance = tf.Print(variance, [tf.sqrt(variance + epsilon)], message='sqrt(variance(x)): ')
variance = tf.Print(variance, [tf.sqrt(avg**2 + epsilon)], message='sqrt(mean(x)^2): ')

x = (x - avg) / (tf.sqrt(variance + epsilon))

# Run
sess = tf.InteractiveSession()
sess.run(tf.global_variables_initializer())
sess.run(x, feed_dict={o: np.random.rand(32, 84, 84, 4)})
```

And the output:

mean(x): [[[[0.176351935 -0.11745432 -0.17503795]]]...]
variance(x): [[[[0.0344150327 0.0378020629 0.0353332]]]...]
sqrt(variance(x)): [[[[0 0 0]]]...]
sqrt(mean(x)^2): [[[[0 0 0]]]...]

**Comment by wei-v-wang** [comment]
Thank you Victor @victorcampos7 for trying Tensorflow with MKL-DNN and for filing this detailed, easy to reproduce bug report. Please know that I have found at least this TF commit id (TF1.5, v1.5.0-2054-gabccb5d3cb ) and this TF commit id (v1.6.0-rc1-1587-g10cddb4268)
produced non-zero results. 

Log is below: (for the v1.5 one)
mean(x): [[[[-0.105329745 -0.242625624 -0.45726034]]]...]
variance(x): [[[[0.02950803 0.0405935161 0.0426237248]]]...]
sqrt(variance(x)): [[[[0.174665481 0.203944862 0.208862931]]]...]
sqrt(mean(x)^2): [[[[0.109974332 0.244677722 0.458352506]]]...]

for the v1.6 one:
mean(x): [[[[-0.253690451 0.0252816677 -0.101222448]]]...]
variance(x): [[[[0.0222420059 0.0317621827 0.0308733471]]]...]
sqrt(variance(x)): [[[[0.152453288 0.181003273 0.178531066]]]...]
sqrt(mean(x)^2): [[[[0.255653769 0.0404865742 0.106047086]]]...]

I suggest trying this TF commit id: abccb5d3cb45da0d8703b526776883df2f575c87 

Please let us know if you still have issues!

**Comment by wei-v-wang** [comment]
Also, below is from GPU TF as a FYI

mean(x): [[[[-0.152227402 -0.730839372 -0.154905677]]]...]
variance(x): [[[[0.0293967575 0.0375702456 0.0372712091]]]...]
sqrt(variance(x)): [[[[0.174346656 0.196393088 0.195630282]]]...]
sqrt(mean(x)^2): [[[[0.15547727 0.731523156 0.158100501]]]...]

**Comment by victorcampos7** [comment]
Thanks for the feedback. We installed TF 1.6.0 (the version in the releases site) which already includes the commits you mention, but the error persists. Does it work properly for you in the stable version, or only in v1.6.0-rc1-1587-g10cddb4268?

**Comment by tripiana** [comment]
Should it be. I've double checked this commit, and it's applied in the v1.6.0 tag.

@wei-v-wang @skye any ideas?

**Comment by wei-v-wang** [comment]
Read your post carefully, you indeed mentioned that on machines that do not support AVX512, things seemed to work. 
I think there is a known issue for Eigen AVX512 and we/Intel are working with Eigen community to fix that. https://bitbucket.org/eigen/eigen/pull-requests/361/fix-for-avx512/diff 
So the workaround is please for now build with "-march=broadwell" before Eigen AVX512 issue is fully fixed.

**Comment by tripiana** [comment]
Our native is skylake avx512.

Carlos.

**Comment by tripiana** [comment]
We use copt to add march mtune flags

Carlos.

**Comment by wei-v-wang** [comment]
Please use -march=broadwell (for now), both from the ./configure stage, and the bazel build stage.

**Comment by wei-v-wang** [comment]
In the meantime, I will double check with stable r1.6 and -march=native -mtune=native. Thanks!

**Comment by wei-v-wang** [comment]
I confirm with "-march=native -mtune=native", due to Eigen AVX512 failure (which we are fixing), will reproduce the bug. 
So, the current workaround is to please temporarily use "-march=broadwell" (i.e. for Eigen pass, do not use AVX512). Please be assured that when code is executing in MKL-DNN pass, AVX512 will be used automatically if SKX is used.

**Comment by wei-v-wang** [comment]
Please make sure to specify during the configure step. specifying in bazel build is not enough. 
Please specify optimization flags to use during compilation when bazel option "--config=opt" is specified [Default is -march=native]: -march=broadwell

**Comment by tripiana** [comment]
I can confirm that our build with broadwell works as expected and the bug do no appear. Will it be superb to post here when TF will be working with avx512 again. Thanks for the support.

**Comment by wei-v-wang** [comment]
Thanks @tripiana for trying with the workaround and your suggestion to post updates! Yes, definitely. I am planning to post here when there is good news.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by wei-v-wang** [comment]
Eigen community included our fix for sqrt in the following commit https://mirror.bazel.build/bitbucket.org/eigen/eigen/get/87636bcbdf12.tar.gz 

We have created and are testing a PR to tensorflow to use the above Eigen to fix avx512 sqrt/rsqrt issue.

**Comment by pierremac** [comment]
I'm facing a similar problem with batch_normalization on TF 1.7 compiled with support of AVX512.
I'm not even using tf.moments and instead set the variance as ones.
When I pass it to the batch_normalization (along with any kind of variance_epsilon), the rsqrt operation (in the computation of "inv") returns NaN.
Everything works correctly with the default pip tensorflow-gpu 1.7 package (which does not support AVX512)

**Comment by wei-v-wang** [comment]
https://github.com/tensorflow/tensorflow/pull/18411 includes fix that will address Eigen rsqrt AVX512 issue. 
However, we are working on a complete solution to fully enable AVX512. More updates will be provided.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by skye** [comment]
@wei-v-wang any update on this?

**Comment by wei-v-wang** [comment]
@skye we are still working on this. Looks like https://github.com/tensorflow/tensorflow/pull/18411 is having some conflicts that we need to address first.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by wei-v-wang** [comment]
Pending on the status of #18411

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 30 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
We are closing this issue for now due to lack of activity. Please comment if this is still an issue for you. Thanks!
